lib/tree.rb

```diff
@@ -16,42 +16,128 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n) for balanced trees, O(n) for unbalanced trees.
+  # Space Complexity: O(1)
```

Comment:
Since you have a recursive `add_helper` you are using the system stack and making O(log n) recursive calls which go on the stack.  So you have O(log n) space complexity.

This applies to your other recursive methods as well.